src/main/java/bio/terra/cloudfunctions/common/FileTypeDetector.java

```diff
@@ -0,0 +1,80 @@
+package bio.terra.cloudfunctions.common;
+
+import bio.terra.cloudfunctions.utils.MediaTypeUtils;
+import com.google.events.cloud.storage.v1.StorageObjectData;
+import java.io.BufferedInputStream;
+import java.io.InputStream;
+import java.util.logging.Logger;
+
+/**
+ * This class detects the raw file type from the StorageObjectData and opens an appropriate
+ * InputStream to the underlying Google storage object.
+ *
+ * <p>Sub-classes of this class can be integrated with a DI framework to deploy Cloud Functions like
+ * a Service.
+ */
+public class FileTypeDetector {
+  private static final Logger logger = Logger.getLogger(FileTypeDetector.class.getName());
+  private StorageObjectData storageObjectData;
+  protected InputStream inputStream;
+  protected InputStream dataStream;
+
+  public FileTypeDetector(StorageObjectData storageObjectData) {
+    this.storageObjectData = storageObjectData;
+  }
+
+  private MediaTypeWrapper getMediaType() {
+    return new MediaTypeWrapper(storageObjectData.getContentType());
+  }
+
+  public InputStream getDataStream() {
+    return dataStream;
+  }
+
+  public void setInputStream(InputStream inputStream) {
+    this.inputStream = inputStream;
+  }
+
+  /**
+   * Obtain an input stream from the GCS object and convert it to a typed input stream. It supports
+   * GZIP and JSON file types. For GZIP, the method returns the ArchiveInputStream or
+   * BufferedInputStream to dataStream property, depending on whether the GZIp is a TAR archive or
+   * not. If the file type is JSON, then BufferedInputStream will be returned to dataStream.
+   */
+  public void handleMediaType() throws Exception {
+    String projectId = System.getenv("GCLOUD_PROJECT");
+    if (inputStream == null)
+      inputStream =
+          MediaTypeUtils.getStorageObjectDataAsInputStream(
+              projectId, storageObjectData.getBucket(), storageObjectData.getName());
+    if (getMediaType().isApplicationGzip()) {
+      dataStream = handleGzipType(inputStream);
+    } else if (getMediaType().isApplicationJson()) {
+      dataStream = handleJsonType(inputStream);
+    }
+  }
```

Comment:
I think that simply looking at the `Content-Type` of the incoming file is restrictive and may cause us headaches in the future. For instance, what if QA and App Services both want to process some json, but using different json schemas? Both of these would show a content type of `application/json`, but they would need different handlers.

I'd prefer one of two routes: 
1) give as much information to the switching logic as possible - this would include the file path and bucket, the content-type, the triggering event type, really anything we can think of. Then, developers can write flexible switching logic such as "if it is json, inside the 'foo' bucket, and with an '.upsert' extension, use class Y for processing".
2) specify the handler class in a config file. App Services, QA, and future callers can deploy their instance and specify their main handler class in config; at runtime, we'll simply use that handler class and avoid switching logic. If a given deployment needs custom switching logic, they can write it themselves in their handler class.